pkg/controller/cainjector/controller.go

```diff
@@ -0,0 +1,211 @@
+	// CAReferences returns the set of url or service references
+	// in the given target that can have CA data (e.g. APIService
+	// has CAData, admission webhooks have one ClientConfig in
+	// each sub-webhook)
+	CAReferences() []Injectable
+}
+
+// Injectable is a point in a Kubernetes API object that represents a Kubernetes Service
+// reference with a corresponding spot for a CA bundle.
+type Injectable interface {
+	// SetCA sets the CA of this injectable to the given certificate data.
+	SetCA(data []byte)
+}
+
+// CertInjector knows how to create an instance of an InjectTarget for some particular type
+// of inject target.  For instance, an implementation might create a InjectTarget
+// containing an empty MutatingWebhookConfiguration.  The underlying API object can
+// be populated (via AsObject) using client.Client#Get, and then CAs can be injected with
+// Injectables (representing the various individual webhooks in the config) retrieved with
+// Services.
+type CertInjector interface {
+	// NewTarget creates a new InjectTarget containing an empty underlying object.
+	NewTarget() InjectTarget
+}
+
+// genericInjectReconciler is a reconciler that knows how to check if a given object is
+// marked as requiring a CA, chase down the corresponding Service, Certificate, Secret, and
+// inject that into the object.
+type genericInjectReconciler struct {
+	// injector is responsible for the logic of actually setting a CA -- it's the component
+	// that contains type-specific logic.
+	injector CertInjector
+
+	log logr.Logger
+	client.Client
+
+	resourceName string // just used for logging
+}
+
+// InjectClient provides a client for this reconciler.
+func (r *genericInjectReconciler) InjectClient(c client.Client) error {
+	r.Client = c
+	return nil
+}
+
+// splitNamespacedName turns the string form of a namespaced name
+// (<namespace>/<name>) back into a types.NamespacedName.
+func splitNamespacedName(nameStr string) types.NamespacedName {
+	splitPoint := strings.IndexRune(nameStr, types.Separator)
+	if splitPoint == -1 {
+		return types.NamespacedName{Name: nameStr}
+	}
+	return types.NamespacedName{Namespace: nameStr[:splitPoint], Name: nameStr[splitPoint+1:]}
+}
+
+// Reconcile attempts to ensure that a particular object has all the CAs injected that
+// it has requested.
+func (r *genericInjectReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.log.WithValues(r.resourceName, req.NamespacedName)
+
+	// fetch the target object
+	target := r.injector.NewTarget()
+	if err := r.Client.Get(ctx, req.NamespacedName, target.AsObject()); err != nil {
+		log.Error(err, "unable to fetch target object to inject into")
+		return ctrl.Result{}, err
+	}
+
+	// ensure that it wants injection
+	metaObj, err := meta.Accessor(target.AsObject())
+	if err != nil {
+		log.Error(err, "unable to get metadata for object")
+		return ctrl.Result{}, err
+	}
+	certNameRaw := metaObj.GetAnnotations()[WantInjectAnnotation]
```

Comment:
Can `metaObj.GetAnnotations()` return nil?

Reply:
sure, but it doesn't matter.  Indexing into a nil map is an acceptable operation in Go (https://golang.org/ref/spec#Index_expressions) and just returns the zero value (an empty string)